presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java

```diff
@@ -936,7 +941,33 @@ public Object visitCast(Cast node, Object context)
         @Override
         protected Object visitArrayConstructor(ArrayConstructor node, Object context)
         {
-            return visitFunctionCall(new FunctionCall(QualifiedName.of(ArrayConstructor.ARRAY_CONSTRUCTOR), node.getValues()), context);
+            ImmutableSet.Builder<Type> elementTypes = ImmutableSet.builder();
+            List<Object> arrayValues = new ArrayList<>();
+            boolean containsExpression = false;
+            for (Expression expression : node.getValues()) {
+                Object value = process(expression, context);
+                if (value instanceof Expression) {
+                    containsExpression = true;
+                    break;
+                }
+                Type type = expressionTypes.get(expression);
+                arrayValues.add(value);
+                elementTypes.add(type);
+            }
+
+            if (containsExpression || arrayValues.isEmpty()) {
```

Comment:
Now that you can tell the `elementType` even with empty array constructor, you don't need the second parameter any more.